### DISCUSSION TITLE
"CornerRadius" is a misunderstanding.

### DISCUSSION THREAD

**Discussion by Splitwirez** [AvaloniaUI/Avalonia#5594] [Ideas]
Hi,

Recently, while working on a [PR to port the WinUI `Expander` style to Avalonia](https://github.com/AvaloniaUI/Avalonia/pull/5573), I ran into some trouble stemming from how WinUI has a `CornerRadius` property on most controls, while Avalonia does not, and was reminded of an observation I made a while ago.

Metaphorically speaking, a `CornerRadius` property on all applicable controls (as is the case with `BorderThickness`) would ask the app developer a question:
* What should the radius of each corner of this control be?

But I feel that this is an unfortunate, perhaps-misguided conflation of  *two* questions, which should be able to be answered independently of one another:
1. Which of these corners should be rounded?
2. What is the radius of a rounded corner?

This distinction is important. For example, take this example:
![AvaloniaCornersExample](https://user-images.githubusercontent.com/34009058/109704690-e55dae00-7b64-11eb-9bad-84ec06970976.png)
Here, we see three linked buttons, with only the left corners of the leftmost button and the right corners of the rightmost button being rounded. Currently, there is no built-in way to do this in Avalonia, WinUI, WPF, or any other XAML framework I know of, without forcibly answering both of the above questions. Avalonia's `/template/` selector element reduces the amount of redundancy caused by this, but still doesn't solve the underlying problem.

This inability to answer one question without answering the other means a lot of hardcoded corner radii. These can and will later prove to be quite the nuisance, should the developer later decide to switch their app over to a new theme, or to provide multiple themes for the user to choose from, or if the app uses say, Avalonia's Fluent theme, and an Avalonia update changes the default corner radius for the Fluent theme, or any number of other such scenarios.

So that's the problem. Now, onto what I propose could be done about it.

For a Control library I've been working on, I experimented with a solution to this, which would allow these two questions to be answered independently of each other, as separate properties. That solution was as follows:

* A new class similar to `CornerRadius` but with four `bool`s instead of four `double`s, entitled `CornerCurves`
* An attached property for all applicable controls, called `CornerCurves` (of type `CornerCurves`, predictably enough)
* A value converter which takes in the attached `CornerCurves` property and a `CornerRadius` object, and produces a new `CornerRadius` object, with each corner's radius being either that of the original input `CornerRadius` if the `CornerCurves` object had `true` for that corner, or zero if the `CornerCurves` object had `false` for that corner

My exact implementation of this solution was flawed in various ways, but it *did* solve the problem: It allowed the app developer to answer *only* the first question ("Which of these corners should be rounded?"), while leaving the Avalonia theme still able to answer the second question ("What is the radius of a rounded corner?"), with no additional action from the developer being required.

I believe a similar solution can and should be implemented into Avalonia (albeit presumably *not* as an *Attached* property). I'm eager to do my part to do so, but want to hear what others think about both the problem and the proposed solution.

~~Also this would need a way to pass the value of a `DynamicResource` into a converter, or failing that a `StaticResource`...neither of which I've figured out a way to achieve, thus far...~~

---

**UPDATE: Following the discussion which took place in the responses to this conversation, I've come to realize that, while this _is_ a real and valid problem, it's not nearly as big of a deal to most as I expected, to the point where a solution to it - or at least a solution as drastic as the one I proposed - would be more trouble than it's worth to most.**

**Maybe Avalonia _should_ just follow WinUI's example here. If that ends up being the plan, I'd be more than happy to do what I can to help make it happen.**

**Comment by robloo** [comment]
Conceptually I think the answer here is simple. CornerRadius should be alongside BorderThickness property. In WPF/UWP BorderThickness is not part of UIElement or FrameworkElement, it is part of Border and Individual controls have this property themselves. I don't think its in all controls (even the ones that don't make sense) as you imply. Avalonia should do the same and only add it to Controls where it makes sense.

About the separation of concepts: only controls where a corner radius makes sense have the property. Setting any radius to zero turns it off. All information is clear. Your assumption that all controls have a corner radius property even if it doesn't make sense is incorrect I think. The existence of the property itself fills in the missing information you are looking for.

**Reply by Splitwirez** [reply]
> About the separation of concepts: only controls where a corner radius makes sense have the property.

My bad, I meant all controls where the property has an actual meaning. There, I've updated the original post accordingly.

> Conceptually I think the answer here is simple. CornerRadius should be alongside BorderThickness property.

...that's literally what I'm saying I *don't* think should be done. In fact, I'd argue that `BorderThickness` has the same problem.

**Reply by robloo** [reply]
BorderThickness like BorderBrush makes sense in a lot of scenarios though. It's very common in styling and it just needs to be included in the controls as a property to make development/styling easier. Fill/Stroke is a fundamental concept for graphical objects as well. (Fill=Background, Stroke=BorderBrush (and with BorderBrush has to go BorderThickness)). This all makes perfect sense to me and is so deeply ingrained in not only WPF but other graphical systems I think it would be a mistake to consider changing it. With BorderThickness established CornerRadius is established as well.

**Reply by Splitwirez** [reply]
I'm not sure I agree with that.

Another example of where this comes into play is if you have, say, a card or popup or something like that with rounded corners, and want to have, for example, a footer, which has those bottom corners also rounded. *Can* you do this with `CornerRadius`? Yes. But you can't do it without locking down the answers to both questions, as I explained before.

**Reply by robloo** [reply]
Well, I'm not sure we will agree here :) You are leading more towards a group of controls concept here as well so I'll just talk about that below.

**Reply by Splitwirez** [reply]
> Well, I'm not sure we will agree here :)

You're right. But I think you probably represent the majority better than I do. I've revised the OP to reflect this realization on my part.

**Reply by maxkatz6** [reply]
@Splitwirez talking about your example with Popup+Footer, I can try to answer your questions from my point of view:

### > "What is the radius of a rounded corner?"

It have to use same radius as a popup, preferably with xaml-friendly syntax like Bindings. This part already easily possible in any popular XAML framework. What is not defined in xaml "standard" is where you can get this radius.
In UWP there is CornerRadius for controls where it is needed, this is preferable way as for me, and I agree with @robloo here.
In older UWP versions and WPF/Avalonia you need to bind directly to the Popup's Border control (if possible) or use shared resource (as it was done in WinUI before).


### > "Which of these corners should be rounded?"

This question is a bit more tricky. I see next possible answers:

1. Right now it is possible with converter which was discussed in Expander control:
```xaml
<CornerRadiusFilterConverter x:Key="TopCornerRadiusFilterConverter" Filter="Top"/>
...
{Binding CornerRadius, RelativeSource={RelativeSource TemplatedParent}, Converter={StaticResource TopCornerRadiusFilterConverter}}
or
{Binding Source={StaticResource OverlayCornerRadius}, Converter={StaticResource TopCornerRadiusFilterConverter}}
```
This approach arguably is kind of verbose, but it's XAML-friendly. WinUI also has predefined converters [in the resources](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/CommonStyles/CornerRadius_themeresources.xaml).
Also approach with shared resource is not so usable in WPF/Avalonia, because it has to be StaticResource, and DynamicResource can't use converters. Although it's possible to define some invisible proxy control with CornerRadius from dynamic resource and bind to it with Binding+Converter, but it's **ugly solution**.

2. Your specific Popup+Border problem can be also solved with ClipToBounds="True" on outer border with corner radius. But it's not that flexible and doesn't solve cases when you want to have some padding between borders.

3. I would love to see ability to define thickness and cornerradius properties without need to define all corners/sides at once. It is possible to do so with special attached properties now, but I it's not out-of-box feature.
```xaml
<Style Selector="Border.inner">
   <!-- Meanwhile we have same problem with DynamicResource here -->
   <Setter Property="CornerRadius.BottomLeft" Value="{Binding Path=BottomLeft, Source={StaticResource OverlayCornerRadius}}" />
   <Setter Property="CornerRadius.BottomRight" Value="{Binding Path=BottomRight, Source={StaticResource OverlayCornerRadius}}" />
</Style>
```

**Reply by maxkatz6** [reply]
Continuing answering your second question, I want to say, that it will be better to implement ability to use converters with dynamic resources OR/AND allow partial structure updating (it's obviously will require recreating whole struct in the IL code, but this part can be hidden from the user). @grokys @kekekeks 

Before that I am pretty fine with porting WinUI/UWP CornerRadius properties for some controls. @danwalmsley

**Reply by Splitwirez** [reply]
I just thought of another idea: Take the approach I described initially, with the `CornerCurves` property, but instead of using a converter to combine them, add a `CornerCurves` property to the `Border` control, `TemplateBind` it in the template, and have the `Border` control itself do all the deciding-which-corners-to-round.

This would allow for `DynamicResource`s to be used, while still allowing us to reap the benefits I initially highlighted.

**Reply by Splitwirez** [reply]
@maxkatz6 tbh, I feel like I've not done a very good job of explaining what I'm suggesting. I'm thinking it might be easier if I were to put together a draft PR to add the necessary plumbing for one or two controls to show how it could work, and then I would either finish the job if you folks give the go-ahead, or scrap it and help add the `CornerRadius` properties if you folks reject it.

Does that make any sense to you?

**Reply by maxkatz6** [reply]
> Take the approach I described initially, with the CornerCurves property, but instead of using a converter to combine them, add a CornerCurves property to the Border control

I understand this approach, but for me it looks like a workaround build in the framework. For user it might be unintuitive why he has two APIs to choose what corners to display: CornerRadius and CornerCurves. It makes less sense if we will describe it as "CornerCurves used only in some specific use cases, when we bind to CornerRadius and our converters/setter is not good enough to make it simple".

**Reply by maxkatz6** [reply]
Assuming that we can't improve our Binding/DynamicResource integration or/and improve setters syntax in near time, I would say that I see adding CornerRadius property to every relevant control and using special converter to select corners (like it's done in WinUI) is more acceptable. Though I would like to hear opinion of other maintainers.

**Reply by maxkatz6** [reply]
But another issue with Bindings to CornerRadius properties is that we can't define that binding inside of template because of priorities issue:
CornerRadius="{Binding CornerRadius, RelativeSource={RelativeSourceTemplateBinding}, Converter={...}}"
will have Local priority value, that can't be overridden by user with his custom style. So we required to define this binding inside of Setter to have Style value, that can be overridden.

**Comment by robloo** [comment]
I should also add, I think the group of buttons you mentioned is an interesting example. This ideally should be a separate control (like ButtonGroup/SegmentedControl [WinUI discussion](https://github.com/microsoft/microsoft-ui-xaml/issues/2310)or the [RadioButtons group control](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.radiobuttons?view=winui-2.5)). However, this could be generalized and a 'GroupedContainer' parent control could be created. This grouped container would take care of managing border/corner radius of children along with orientation. The container would have the properties that would apply to the group as a whole.

**Reply by Splitwirez** [reply]
That's actually an interesting idea, now that you  mention it.

Question though...how would this hypothetical "GroupedContainer" parent control declare whether or not each of its childrens' corners should be rounded without locking down how round "round" corners should be?

And what is one to do when the controls are being placed adjacent to one another in, say, a grid-based fashion? Is that somehow also the domain of the GroupedContainer, or...?

**Reply by robloo** [reply]
The easiest method would be to have two CornerRadius properties `CornerRadius` for the group edges and `ContentCornerRadius` or similar for content controls themselves. However, the best action by far for the example you gave is to create a new ButtonGroups control (support both single/multi-selection).

**Reply by Splitwirez** [reply]
> The easiest method would be to have two CornerRadius properties `CornerRadius` for the group edges and `ContentCornerRadius` or similar for content controls themselves.

Okay, that...that makes sense.

> However, the best action by far for the example you gave is to create a new ButtonGroups control (support both single/multi-selection).

That's definitely *also* an interesting idea...though it doesn't exactly solve the underlying problem.

These are good ideas, but they're all trying to solve one specific use case, rather than addressing the actual underlying problem.

**Reply by robloo** [reply]
I think fundamentally, a more abstract 'group of controls' concepts is needed to address all your concerns. Conceptually, it isn't just CornerRadius to talk about here as you pointed out. Other types of properties could run into this situation (BorderThickness is a common one). Even something like sharing the Background between controls is relevant in the generalized discussion.

A hypothetical GroupContainer control would have it's own properties and logic (something like converters) determining how to apply properties to children. However, it isn't only styling concerns that would need to be addressed. Actual logic such as the segmented control example would need to be implemented as well. The group needs to be aware of pointer over, clicking, selection, etc. How is it possible to generalize this? This high-level of abstraction is probably more trouble than it's worth (even if it's possible). Perhaps making a 'StyleGroupContainer' is possible to make it clear logic must be handled elsewhere.

Anyway, I know I'm focusing on more than just CornerRadius but that seems to be unavoidable with the points you raise. At the end of the day for me the current level of abstraction we have now makes the most sense. Each control needs to manage things themselves but they also need to share as many properties as possible to make styling consistent (and sane). For the common/special cases RadioButtonGroup/ButtonGroup controls should be created.

Also, you are essentially questioning the original decision to add Border/Background properties to controls themselves. Again, this is just development convenience (and also was backwards compatible with Windows Forms at the time). There is no limit to how far this line of thinking can go though. There is constantly a struggle between optimizing ease-of-use with general-purpose code. Each control could have have a property for every single thing in the style that is configurable but that just doesn't make sense -- it also makes every control so different it's almost unusable. Things simply evolved to some point in the middle and the current level of abstraction seems to work for the vast majority of developers. I guess I just don't share your concerns overall.

**Reply by Splitwirez** [reply]
Funny you'd mention backgrounds...I'd previously thought about the potential value of being able to have, say, a gradient brush, which is drawn to an area relative to one control, but whose stops are positioned relative to another.

Who knew control properties could be so...uh..."philosophical" isn't quite the right word...eh, something. Not sure what, but definitely something.


~~Also tbh, I may or may not be slightly biased, due to the `CornerRadius`/`BorderThickness` case in particular being so relevant specifically to user-skinnable apps~~

**Reply by maxkatz6** [reply]
RadioButtons/GroupedContainer can be useful control that solves some specific problem, but it won't solve another cases like Popup+Border example from another thread. Let's say we have inner and other Borders, where we want to reuse same CornerRadius, but with different Top values and without defining hardcoded values in the resources (which can be fine in developer's applications, but won't be fine in the control libraries). I wrote some my thought in that another thread as well.

![image](https://user-images.githubusercontent.com/3163374/110189852-af335f00-7dee-11eb-9c91-177789215768.png)

**Reply by robloo** [reply]
In your example, Popup would have a CornerRadius StyledProperty that binds to all elements with a corner radius. The filter converter for corner radius would be used on the inside border that doesn't have rounded corners. I think this could be handled just like WPF/WinUI. Do I misunderstand the concerns?

**Reply by maxkatz6** [reply]
@robloo you are right, that's what I meant. Converter solves this problem, but it doesn't work well with shared resources, as I described here https://github.com/AvaloniaUI/Avalonia/discussions/5594#discussioncomment-434984 (if we don't have CornerRadius property). 
And also Popup probably shouldn't have CornerRadius property anyway, because it's more low-level control that can't map this radius to native components. Although Flyout potentially should have it.

**Comment by Sorien** [comment]
maybe the whole problem is that Avalonia is missing first-child and last-child selector https://quirksmode.org/css/selectors/firstchild.html

edit:
some css sample  https://jsfiddle.net/ghc1v3ye/

**Reply by Splitwirez** [reply]
Like with @robloo 's second post, this would only solve one specific use case, rather than addressing the actual underlying problem. ~~Still probably a good thing to have though~~

**Reply by robloo** [reply]
This is actually a good idea I think. It fits in with the concepts as they stand today as well.

**Reply by Splitwirez** [reply]
Agreed

**Reply by maxkatz6** [reply]
I wonder, should these pseudoclasses work with materialized containers for Items or should it detect first/last child from the ItemsLayout/ItemsPanel.
But in any way this feature will benefit in most of use cases.
